### PR TITLE
[List v2]: Handle Enter in empty list items

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/39519
Alternative of: https://github.com/WordPress/gutenberg/pull/39626
Based on @mtias's suggested algorithm here: https://github.com/WordPress/gutenberg/pull/39626#issuecomment-1080555671 

>I think pressing `enter` on an empty list item should backtrack one level at a time. This video is from Apple Notes as an example (I'm only pressing enter):

https://user-images.githubusercontent.com/548849/160392233-d4345663-f3ba-491d-ab7f-1b96456d90e8.mov

In the experimental new list blocks we should split lists on empty list items if we press enter.

## How?
Since my first approach with a different algorithm, the indentation has been implemented by @jorgefilipecosta and I'm using one of the hooks for one of the two parts of the algorithm. 

## Testing Instructions
1. Insert a new list block with any nested lists
2. On an empty list item press `Enter`
3. Should split the list and insert a paragraph in the middle
4. If the list item we press Enter is nested it should outdent

## Screenshots or screencast 

https://user-images.githubusercontent.com/16275880/160637870-3663d9a3-321d-41f7-aced-ab89ee6e0ac6.mov

## Notes
1. The handling of `Enter` in a non empty list item is not handled yet, as it needs more changes involving the RichText API.
2. I've created a new `hooks` file and moved `useIndentListItem and useOutdentListItem` without making any change. The new hook is the `useEnter` one.

### CODE REVIEW

File: packages/block-library/src/list-item/hooks.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+				selectionEnd.clientId === selectionStart.clientId
+					? selectionStart.offset
+					: selectionEnd.offset,
+				selectionEnd.offset
+			);
+		}, [ clientId ] ),
+	];
+}
+
+export function useEnter( props ) {
```
Comment (ellatrix) [inline packages/block-library/src/list-item/hooks.js]: Can we separate these functions in separate files?

**Comment by ellatrix** [comment]
I like this behaviour :)

File: packages/block-library/src/list-item/hooks.js (hunk 1 above)
Comment (ntsekouras) [reply packages/block-library/src/list-item/hooks.js]: Done.

File: packages/block-library/src/list-item/hooks/use-enter.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+				if ( event.defaultPrevented || event.keyCode !== ENTER ) {
+					return;
+				}
+				const { content, clientId } = propsRef.current;
+				if ( content.length ) {
+					return;
+				}
+				event.preventDefault();
+				if ( canOutdent ) {
+					outdentListItem();
```
Comment (ellatrix) [inline packages/block-library/src/list-item/hooks/use-enter.js:45]: Shouldn't this be indent?

**Review by ellatrix — APPROVED** [review]
Looks good to me, but the behaviour seems to be the opposite?

**Comment by ellatrix** [comment]
@ntsekouras In a separate PR, can you do the same thing for `Backspace`? Gradually the list item? Worth checking how it works in list v1, I think the behaviour there is good.

File: packages/block-library/src/list-item/hooks/use-enter.js (hunk 2 above)
Comment (ntsekouras) [reply packages/block-library/src/list-item/hooks/use-enter.js:45]: Why? It's working as expected in the video, no?

**Comment by mtias** [comment]
Sweet! :)

File: packages/block-library/src/list-item/hooks/use-enter.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-library/src/list-item/hooks/use-enter.js:45]: For me it was indenting for some reason instead of outdenting.

**Pull request merged**